text/0000-safe-unsafe-trait-methods.md

```diff
@@ -0,0 +1,232 @@
+- Feature Name: safe_unsafe_trait_methods
+- Start Date: 2018-01-30
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This RFC allows safe implementations of `unsafe` trait methods.
+An `impl` may implement a trait with methods marked as `unsafe` without
+marking the methods in the `impl` as `unsafe`. This is referred to as
+*overconstraining* the method in the `impl`. When the trait's `unsafe`
+method is called on a specific type where the method is known to be safe,
+that call does not require an `unsafe` block.
+
+# Motivation
```

Comment:
Do you have some real world examples of traits and implementations that would take advantage of this?

Reply:
Here's one example: https://doc.rust-lang.org/std/slice/trait.SliceIndex.html#tymethod.get_unchecked
with impls:
+ https://doc.rust-lang.org/src/core/slice/mod.rs.html#967-999
+ https://doc.rust-lang.org/src/core/str/mod.rs.html#1827-1853

While this falls under "trivial example", avoiding `unsafe` where possible is still nice =)

...stay tuned for more

Reply:
Does there exist any code that has ever called `slice.get_unchecked_mut(..)`?

Reply:
Oh; you were referring to the calling side of things and not the motivation, I see. then `slice.get_unchecked_mut(..)` does not apply (I think). My bad ;)

Reply:
I am trying to figure out what problem that currently exists this RFC is proposing to solve. "If this feature existed it would enable me to make a thing that was previously impossible", or "If this feature existed I could make this thing I wrote better".

Reply:
@sfackler This RFC was motivated in part by the upcoming changes to `futures-rs` that will allow immovable future types. In short, the original plan was to land something like this:

```rust
trait Future {
    trait Item;
    trait Error;
    unsafe fn poll_immovable(&mut self) -> Poll<Self::Item, Self::Error>;
}

trait MovableFuture: Future {
    fn poll(&mut self) -> Poll<Self::Item, Self::Error>;
}
```

This RFC would make it possible to implement `Future` without writing `unsafe`, which makes it easier to validate that `Future` impls are sound. The `Future` trait could also offer a custom-derive `#[derive(MovableFuture)]` which would expand to this:

```rust
impl MovableFuture for MyType {
    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        self.poll_immovable()
    |
}
```

Note that the call to `poll_immovable` is not wrapped in an `unsafe` block. This will produce a compiler error iff the impl of `poll_immovable` is `unsafe`, but will work just fine for safe impls.

In @withoutboats's blog series, they and @eddyb discovered that `fn anchored_poll(self: Anchor<&mut Self>)` could be used to avoid the `unsafe poll_immovable`. Unfortunately, `anchored_poll` isn't object-safe (because of `arbitrary_self_types`), so we still need an underlying `poll_immovable` function under the hood.

There was also a suggestion to use `trait ImmovableFuture` and `trait Future` with `impl<T> ImmovableFuture for T where T: Future`. Unfortunately, this approach requires specialization in order to not conflict with impls like this:

```rust
struct MyFutureCombinator<T> { .... }

impl<T> Future for MyFutureCombinator<T> where T: Future { ... }
impl<T> ImmovableFuture for MyFutureCombinator<T> where T: ImmovableFuture { ... }
```

This all brings us back to `trait Future` and `trait MovableFuture: Future`, where `Future::poll_immovable` is `unsafe`. We want to be able to implement the trait safely, and this RFC would make that possible.

Reply:
> Unfortunately, anchored_poll isn't object-safe (because of `arbitrary_self_types`)

This is not fundamental, it just needs a few more conditions - I've discussed it with @nikomatsakis before and he thinks it's possible.

Reply:
It seems like using specialization to have a `default impl<T> Future for T where T: MovableFuture` would be a better solution to that problem.

Reply:
I think I agree with you that specialization is a better long-term solution.

Reply:
Why is adding a new unstable language feature a shorter term solution that using another unstable language feature?

Reply:
@sfackler Perhaps I'm using it wrong, but I don't believe the current version of specialization allows this specific impl. The following errors with "conflicting implementations of trait `ImmovableFuture` for type `MyFutureCombinator<_>`":

```rust
#![feature(specialization)]

trait Future {}
trait ImmovableFuture {}

default impl<T> ImmovableFuture for T where T: Future {}

struct MyFutureCombinator<T>(T);
impl<T> Future for MyFutureCombinator<T> where T: Future { }
impl<T> ImmovableFuture for MyFutureCombinator<T> where T: ImmovableFuture { }
```

Reply:
Either way, I'd expect this feature to be much quicker and easier to implement and stabilize than even the most minimal version of specialization.

Reply:
Then this can provide some motivation for finishing specialization! IIRC that kind of implementation was covered in at least some revisions of the specialization RFC.

Are there any more motivating examples for this feature? I am kind of concerned with adding a new language feature that will be used for a single trait.